packages/hyperion-autologging/src/ALInteractableDOMElement.ts

```diff
@@ -44,45 +46,113 @@ type HTMLElementWithHandlers = HTMLElement & {
 }
 
-function eventHandlerTrackerAttribute(eventName: string): string {
-  return `data-${eventName}able`;
+const EventHandlerTrackerAttribute = `data-interactable`;
+
+
+const InteractableAncestor = `interactableAncestor`;
+type InteractableAncestorCache = {
+  [index: string]: (HTMLElement | null)[];
 }
 
-export function getInteractable(
-  node: EventTarget | null,
-  eventName: UIEventConfig['eventName'],
-  returnInteractableNode: boolean = false,
-  // Whether to require an actual handler is assigned to determine interactiveness, rather than including "interactive" element tags
-  requireHandlerAssigned: boolean = false,
-): HTMLElement | null {
-  // https://www.w3.org/TR/2011/WD-html5-20110525/interactive-elements.html
-  const selectorString = `[${eventHandlerTrackerAttribute(eventName)}="1"]${requireHandlerAssigned ? '' : ',input,button,select,option,details,dialog,summary,a[href]'}`;
-  if (node instanceof HTMLElement) {
+let getInteractableImpl: (node: HTMLElement, eventName: UIEventConfig['eventName'], requireHandlerAssigned: boolean) => HTMLElement | null = (node, eventName, requireHandlerAssigned) => {
+  function getInteractableOptimized(node: HTMLElement, eventName: UIEventConfig['eventName'], requireHandlerAssigned: boolean, selectorString?: string): HTMLElement | null {
+    /**
+     * We should be careful to only cache the result based on given arguments. We use a map from eventName to a array based on requiredHandlerAassigned
+     * In this way, each node may point to its closest ancestor that matches the criteria of the interactablity.
+     * The hope is that gradually (specially with mouseover & mousemove) most of this data is calculated and cached, hence reducing the need
+     * for costly computation.
+     */
+    let cached: InteractableAncestorCache | undefined;
+    cached = getVirtualPropertyValue<InteractableAncestorCache>(node, InteractableAncestor);
+    let interactable: HTMLElement | null | undefined = cached?.[eventName]?.[requireHandlerAssigned ? 0 : 1];
+    if (interactable !== void 0) { // Not undefined means we have computed it before
+      return interactable;
+    }
+
+    // https://www.w3.org/TR/2011/WD-html5-20110525/interactive-elements.html
+    selectorString ??= `[${EventHandlerTrackerAttribute}*="${eventName}"]${requireHandlerAssigned ? '' : ',input,button,select,option,details,dialog,summary,a[href]'}`;
+    const element = node;
+    if ((element.matches(selectorString) || elementHasEventHandler(element, eventName as HTMLElementEventNames)) && !ignoreInteractiveElement(element)) {
+      interactable = element;
+    } else if (element.parentElement) {
+      interactable = getInteractableOptimized(element.parentElement, eventName, requireHandlerAssigned, selectorString);
+    } else {
+      interactable = null; // We also cache null to indicate we have already tried for this element. May be unsafe!
```

Comment:
Wouldn't this only be reached if there is no parent,  would this only be a problem for root elements potentially?

Reply:
This is the reimplementation of the loop that we had in the previous version, but turning it into a recursive approach so that each level can check and update the cache. Always after the loop ends we were returning null, so this is going to make sure that's still the case. 
Also in the cache, I'm treating `null` as a valid value since many of the elements along the path from element to root will have root value and we want to skip processing those as well.